core/trino-main/src/main/java/io/trino/sql/planner/optimizations/AddExchanges.java

```diff
@@ -442,6 +442,22 @@ public PlanWithProperties visitTopN(TopNNode node, PreferredProperties preferred
                     return rebaseAndDeriveProperties(node, child);
                 case PARTIAL:
                     child = planChild(node, PreferredProperties.any());
+                    // If source is pre-sorted, partial topN can be replaced with partial limit N.
+                    // We record the pre-sorted symbols in LimitNode to avoid pushdown of such replaced LimitNode
+                    // through the source which was producing ordered input.
+                    List<LocalProperty<Symbol>> desiredProperties = node.getOrderingScheme().toLocalProperties();
+                    if (LocalProperties.match(child.getProperties().getLocalProperties(), desiredProperties).stream()
+                            .noneMatch(Optional::isPresent)) {
```

Comment:
The condition in the `if` is hard to read due to the nesting, chained calls and multi-line expressions. Extract it to a separate variable.

I would restructure as follow:

```java
boolean sortingSatisfied = LocalProperties.match(child.getProperties().getLocalProperties(), node.getOrderingScheme().toLocalProperties()).stream()
        .allMatch(Optional::isEmpty);

if (sortingSatisfied) {
    ....
}
```